task/dump-agent/src/main.rs

```diff
@@ -0,0 +1,219 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Dump Agent
+
+#![no_std]
+#![no_main]
+
+use core::mem::size_of;
+use dump_agent_api::*;
+use idol_runtime::RequestError;
+use static_assertions::const_assert;
+use userlib::*;
+
+//
+// Our DUMP_READ_SIZE must be an even power of 2 -- and practically speaking
+// cannot be more than 1K
+//
+const_assert!(DUMP_READ_SIZE & (DUMP_READ_SIZE - 1) == 0);
+const_assert!(DUMP_READ_SIZE <= 1024);
+
+struct ServerImpl {
+    areas: [DumpArea; 3],
+}
+
+task_slot!(SPROT, sprot);
+
+impl ServerImpl {
+    fn area(&self, index: usize, offset: u32) -> Result<&[u8], DumpAgentError> {
+        if index >= self.areas.len() {
+            Err(DumpAgentError::InvalidArea)
+        } else {
+            let area = self.areas[index];
+
+            if offset < area.length {
+                let addr = (area.address + offset) as *const u8;
+                let len = (area.length - offset) as usize;
+
+                Ok(unsafe { core::slice::from_raw_parts(addr, len) })
+            } else {
+                Err(DumpAgentError::BadOffset)
+            }
+        }
+    }
+
+    fn initialize(&self) {
+        let mut next = 0;
+
+        for area in self.areas.iter().rev() {
+            unsafe {
+                let header = area.address as *mut DumpAreaHeader;
+
+                //
+                // We initialize our dump header with deliberately bad magic
+                // to prevent any dumps until we have everything initialized
+                //
+                (*header) = DumpAreaHeader {
+                    magic: DUMP_UNINITIALIZED,
+                    address: area.address,
+                    nsegments: 0,
+                    written: size_of::<DumpAreaHeader>() as u32,
+                    length: area.length,
+                    agent_version: DUMP_AGENT_VERSION,
+                    dumper_version: DUMPER_NONE,
+                    next,
+                }
+            }
+
+            next = area.address;
+        }
+
+        for area in &self.areas {
+            unsafe {
+                let header = area.address as *mut DumpAreaHeader;
+                (*header).magic = DUMP_MAGIC;
+            }
+        }
+    }
+
+    fn add_dump_segment(&mut self, addr: u32, length: u32) {
+        let area = self.areas[0];
+
+        unsafe {
+            let header = area.address as *mut DumpAreaHeader;
+
+            if (*header).magic != DUMP_MAGIC {
+                panic!("bad dump magic!");
+            }
+
+            let nsegments = (*header).nsegments;
+
+            let offset = size_of::<DumpAreaHeader>()
+                + (nsegments as usize) * size_of::<DumpSegmentHeader>();
```

Comment:
Do we need to check the alignment (in case `sizeof(DumpAreaHeader)` is unaligned w.r.t. `DumpSegmentHeader`) and bounds (in case `nsegments` is large) of `offset`?

Reply:
God, yes. Thank you!